[Media] Adopt WTF::flatMap in CDMInstanceFairPlayStreamingAVFObjC

aestes opened the pull request:
#### e2e4240e63f8dfe5414b9c371c7c7a13a725c254
<pre>
[Media] Adopt WTF::flatMap in CDMInstanceFairPlayStreamingAVFObjC
<a href="https://bugs.webkit.org/show_bug.cgi?id=261794">https://bugs.webkit.org/show_bug.cgi?id=261794</a>
rdar://115758431

Reviewed by Richard Robinson.

Replaced the nested for loops in CDMInstanceSessionFairPlayStreamingAVFObjC::contentKeyGroupDataSourceKeys
with a combination of WTF::flatMap and WTF::compactMap. To support RetainPtr&lt;AVContentKey&gt; in compactMap,
also specialized CompactMapTraits for RetainPtr and added tests.

* Source/WTF/wtf/Vector.h:
(WTF::CompactMapTraits&lt;RetainPtr&lt;T&gt;&gt;::hasValue):
(WTF::CompactMapTraits&lt;RetainPtr&lt;T&gt;&gt;::extractValue):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMInstanceFairPlayStreamingAVFObjC.mm:
(WebCore::CDMInstanceSessionFairPlayStreamingAVFObjC::contentKeyGroupDataSourceKeys const):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/cocoa/VectorCocoa.mm: Added.
(TestWebKitAPI::mapString):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/268208@main">https://commits.webkit.org/268208@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aa50720eb6e4ef14dd3b58b47e203f4af9877ef7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18929 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19270 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19876 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20793 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17697 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19118 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22566 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19397 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19484 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19154 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19265 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16477 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21681 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/16537 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17246 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23664 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/16423 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17513 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17418 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21598 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/18248 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18000 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15268 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/22307 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17080 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5425 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4523 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21441 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/23554 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17836 "Built successfully") | | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5284 "Found 1 jsc stress test failure: microbenchmarks/array-from-object-func.js.lockdown") | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Tools/TestWebKitAPI/Tests/WTF/cocoa/VectorCocoa.mm:
```diff
@@ -0,0 +1,58 @@
+
+namespace TestWebKitAPI {
+
+static RetainPtr<NSString> mapString(RetainPtr<NSString> string)
```
I get why this function exists, but it still looks funny to me 😛

aestes replied:
Fair! I'm using a static function to make it easier to write the `static_assert`; otherwise I'd use a lambda.

rr-codes commented on Tools/TestWebKitAPI/Tests/WTF/cocoa/VectorCocoa.mm:
```diff
@@ -0,0 +1,58 @@
+    @autoreleasepool {
+        Vector<RetainPtr<NSString>> vector { @"one", @"two", nil, @"three" };
+
+        static_assert(std::is_same<decltype(WTF::compactMap(vector, mapString)), typename WTF::Vector<RetainPtr<NSString>>>::value, "WTF::compactMap returns Vector<RetainPtr<>>");
```
why is the `decltype` needed?

aestes replied:
`std::is_same` wants typename arguments, hence `decltype`. What we're asserting is that compactMap returns the expected type given the collection and map function arguments.

aprotyas commented on Tools/TestWebKitAPI/Tests/WTF/cocoa/VectorCocoa.mm:
```diff
@@ -0,0 +1,58 @@
+    @autoreleasepool {
+        Vector<RetainPtr<NSString>> vector { @"one", @"two", nil, @"three" };
+
+        static_assert(std::is_same<decltype(WTF::compactMap(vector, mapString)), typename WTF::Vector<RetainPtr<NSString>>>::value, "WTF::compactMap returns Vector<RetainPtr<>>");
```
Optional nit: why not make the return type assertion with `mapped`? That way you're calling `compactMap()` only once?

```cpp
auto mapped = WTF::compactMap(vector, mapString);

static_assert(std::is_same_v<decltype(mapped), typename WTF::Vector<RetainPtr<NSString>>>, "WTF::compactMap returns Vector<RetainPtr<>>");
```

aestes replied:
I could do that, but to be fair compactMap isn't actually called twice; `decltype` just evaluates what the return type would be given the provided arguments.

aprotyas replied:
oh my bad, ignore

aestes committed:
[Media] Adopt WTF::flatMap in CDMInstanceFairPlayStreamingAVFObjC
https://bugs.webkit.org/show_bug.cgi?id=261794
rdar://115758431

Reviewed by Richard Robinson.

Replaced the nested for loops in CDMInstanceSessionFairPlayStreamingAVFObjC::contentKeyGroupDataSourceKeys
with a combination of WTF::flatMap and WTF::compactMap. To support RetainPtr<AVContentKey> in compactMap,
also specialized CompactMapTraits for RetainPtr and added tests.

* Source/WTF/wtf/Vector.h:
(WTF::CompactMapTraits<RetainPtr<T>>::hasValue):
(WTF::CompactMapTraits<RetainPtr<T>>::extractValue):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMInstanceFairPlayStreamingAVFObjC.mm:
(WebCore::CDMInstanceSessionFairPlayStreamingAVFObjC::contentKeyGroupDataSourceKeys const):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/cocoa/VectorCocoa.mm: Added.
(TestWebKitAPI::mapString):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/268208@main
